pyatv/mrp/__init__.py

```diff
@@ -326,6 +326,68 @@ def artwork_id(self):
         return MrpPlaying(self.psm.playing)
 
 
+class MrpPower(Power):
+    """Implementation of API for retrieving a power state from an Apple TV."""
+
+    def __init__(self, loop, protocol, remote):
+        """Initialize a new MrpPower instance."""
+        super().__init__()
+        self.loop = loop
+        self.protocol = protocol
+        self.remote = remote
+        self._power_state = None
+        self._listener = None
+        self._add_listeners()
+
+    def _add_listeners(self):
+        self.protocol.add_listener(
+            self.update_power_state,
+            protobuf.DEVICE_INFO_MESSAGE)
+        self.protocol.add_listener(
+            self.update_power_state,
+            protobuf.DEVICE_INFO_UPDATE_MESSAGE)
+
+    async def _turn_on(self):
+        await self.remote.menu()
+
+    async def _turn_off(self):
+        await self.remote.home_hold()
+        await self.remote.select()
+
+    async def _get_current_power_state(self):
```

Comment:
You can inline this method.

Reply:
```suggestion
    def _get_current_power_state(self):
```

Reply:
This method is a bit ambivalent. It is named `get_` but also modifies internal state, which is a bit contradictory and I would prefer if you avoid doing that. The power state can be determined by the latest received `DeviceInfoMessage`, which was either the message received when connecting *or* the latest message received in `update_power_state`. Here we can use some python magic...

The `or` operator works with objects. For instance `"hello" or 123` returns `"hello"` and `None or "foobar"` returns `"foobar"`. We can use this logic here to pick which message to use. Lets assume `update_power_state` saves the message internally in `MrpPower` like this:

```python
def __init__(self):
    self.device_info = None  # We must set this in the constructor as well

async def update_power_state(self, message, _):
    self.device_info = message
```

To get the "latest" message, we can now do:
```python
latest_device_info = self.device_info or self.protocol.device_info
```
If we have not received a `DeviceInfoUpdateMessage`, then `self.device_info` will be `None` and we will get the value of `self.protocol.device_info` instead (initial `DeviceInfoMessage`). If we have received the message, then we will get that message instead. So the code would look something like this:

```python
def _get_current_power_state(self):
    latest_device_info = self.device_info or self.protocol.device_info
    return self. _get_power_state(latest_device_info)
```
This has the benefit of being short and does not modify any state and does not require any asyncio operation. You can even inline it into the `power_state` method.

By doing this you don't have to save the state (`_current_power_state`), just keep the `DeviceInfoMessage` instead.